docs/rfc/rfc-004-accounts.md

```diff
@@ -0,0 +1,304 @@
+account to `VestingAccount` and triggering the `TrackDelegation` call. Accounts are already capable of sending messages when
+a state transition, originating from a transaction, is executed.
+
+When accounts receive messages, they will be able to identify the sender of the message and decide how to process the
+state transition, if at all.
+
+### Consequences
+
+These changes would have significant implications for the Cosmos SDK, resulting in a system of actors that are equal from
+the runtime perspective. The runtime would only be responsible for propagating messages between actors and would not
+manage the authorization system. Instead, actors would manage their own authorizations. For instance, there would be no
+need for the `x/auth` module to manage minting or burning of coins permissions, as it would fall within the scope of the
+`x/bank` module.
+
+The key difference between accounts and modules would lie in the origin of the message (state transition). Accounts
+(ExternallyOwnedAccount), which have credentials (e.g., a public/private key pairing), originate state transitions from
+transactions. In contrast, module state transitions do not have authentication credentials backing them and can be 
+caused by two factors: either as a consequence of a state transition coming from a transaction or triggered by a scheduler
+(e.g., the runtime's Begin/EndBlock).
+
+By implementing these proposed changes, the Cosmos SDK will benefit from a more extensible, versatile, and efficient account
+management system that is better suited to address the requirements of the Cosmos ecosystem.
+
+## Implementation
+
+### Account Definition
+
+We define the new `Account` type, which is what an account needs to implement to be treated as such.
+An `Account` type is defined at APP level, so it cannot be dynamically loaded as the chain is running without upgrading the
+node code, unless we create something like a `CosmWasmAccount` which is an account backed by an `x/wasm` contract.
+
+```go
+// Account is what the developer implements to define an account.
+type Account[InitMsg proto.Message] interface {
+	// Init is the function that initialises an account instance of a given kind.
+	// InitMsg is used to initialise the initial state of an account.
+	Init(ctx *Context, msg InitMsg) error
+	// RegisterExecuteHandlers registers an account's execution messages.
+	RegisterExecuteHandlers(executeRouter *ExecuteRouter)
+	// RegisterQueryHandlers registers an account's query messages.
+	RegisterQueryHandlers(queryRouter *QueryRouter)
+	// RegisterMigrationHandlers registers an account's migration messages.
+	RegisterMigrationHandlers(migrationRouter *MigrationRouter)
+}
+
+```
+
+### The InternalAccount definition
+
+The public `Account` interface implementer is then converted by the runtime into an `InternalAccount` implementation,
+which contains all the information and business logic needed to operate the account.
+
+```go
+type Schema struct {
```

Comment:
It's not immediately clear for what and why this is needed. Also, none of these types are defined from what I can tell.

Reply:
@testinginprod could you shed some light on this?

Reply:
Schema is mainly for reflection of messages and state, this will facilitate wallets and explorers (in order to easily discover an account's state).

Reply:
It will also be used to understand if an account adheres to a certain CIP interface, for example account 1234 is a CIP-AbstractedAccount if it accepts MsgAuthenticate message.